flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java

```diff
@@ -1246,8 +1230,18 @@ private Set<String> generateNewTransactionalIds() {
     }
 
     @Override
-    protected void finishRecoveringContext(
-            Collection<FlinkKafkaProducer.KafkaTransactionState> handledTransactions) {
+    protected void finishProcessing(@Nullable KafkaTransactionState transaction) {
+        super.finishProcessing(transaction);
+        // TwoPhaseCommitSink sets transaction = null on final checkpoint and thus closing will leak
+        // the producer. For transactional producers, we track the producer in pendingTransactions.
+        if (transaction != null && !transaction.isTransactional()) {
+            transaction.producer.flush();
+            transaction.producer.close(Duration.ZERO);
+        }
+    }
```

Comment:
I feel uncomfortable merging this change as part of a "fix-tests-ticket". This is a change in production code to more or less support checkpoints-after-task-finish for the deprecated sink function.